MRG: final improvements to connect

jasmainak opened the pull request:
This PR is built on top of #162 so we should wait for that too be merged first. I'm hoping there aren't too many comments on that one :)

To get an idea of the direction take a look at this commit a711e9d

This should truly address #129 and make the code simpler to read. But will need some work to make sure everything is fine.

TODOs

- [x] check pos between unique and common inputs
- [x] tests

jasmainak commented on hnn_core/neuron.py:
```diff
@@ -425,35 +425,93 @@ def _create_cells_and_feeds(self):
                 _PC.cell(gid, feed_cell.nrn_netcon)
 
+    def _connect_celltypes(self, src_type, target_type, loc,
+                           receptor, nc_dict):
```
I think we are getting closer to being able to pass a connectivity matrix ...

ideas on a good/flexible/high-level API are welcome. This was just a first pass. Maybe this could also become a public function if users want to make their own connections.

rythorpe replied:
So the idea here is to make one method of `Network` that calls `parconnect_from_src` and then explicitly define the connections in `Network._parnet_connect()`?

rythorpe replied:
I really like where this is going btw!

jasmainak replied:
yes, the connectivity structure in the old code was really obscure to me. I think it should be transparent and easy to change :) Thanks for the excellent reviews, #162 would not have been possible if we didn't add `sect_loc` attribute also to Basket class and make the calls to `_connect_feed_at_loc` consistent between Pyramidal and Basket class in the previous PR.

jasmainak replied:
`Network._parnet_connect` is the default connectivity that HNN provides but I'm imagining that the user will be able to modify it easily if they want.

jasmainak commented:
@rythorpe @ntolley I did another pass today. I think we are getting closer. However, there is a small bug somewhere due to which the outputs don't match what is on `master`.

If either of you can spot the problem, I buy you a beer next time we meet :) To replicate, just fetch this branch and run:

```py
$ pytest ./hnn_core/tests/test_compare_hnn.py --pdb
```

you can leave out the `--pdb` flag if you don't want to drop into the debugging console when there is an error.

ntolley commented:
I'm down to take a look! Will fetch once I wrap up some tests for the spike rates PR.

Is it a gross mismatch between the two or is the difference more subtle in (floating point error for example).

jasmainak commented:
gross mismatch :-) well, something is off with the scale ... but probably it comes from a tiny bug somewhere.

you can also run [the example](https://github.com/jonescompneurolab/hnn-core/blob/master/examples/plot_simulate_evoked.py). Just stop it at line 56 and change `n_trials=1`. Maybe also change simulation stop time if you want to run faster.

jasmainak commented:
alright, I made it work. Now need to clean up and add tests. Could do with one round of review @rythorpe @ntolley @blakecaldwell

jasmainak commented:
not sure why MPI hates me :-( The MPI test passes locally on my laptop ...

rythorpe commented:
FYI I haven't forgotten about this @jasmainak. I have a chunk of time tomorrow morning scheduled to review your most recent commits.

rythorpe commented on hnn_core/network_builder.py:
```diff
@@ -425,35 +437,184 @@ def _create_cells_and_feeds(self):
+        for gid_target in net.gid_dict[_long_name(target_type)]:
+            if _PC.gid_exists(gid_target):
+                gid_srcs = net.gid_dict[_long_name(src_type)]
+                if unique:
+                    gid_srcs = [gid_target + net.gid_dict[src_type][0]]
```
This feels a bit convoluted. Is this here to assign the subset of source gids that apply to the given target cell type?

jasmainak replied:
see [comment here](https://github.com/jonescompneurolab/hnn-core/pull/164/files#r487466791). The source gids from net.gid_dict are actually ignored if `unique=True`

rythorpe commented on hnn_core/network_builder.py:
```diff
@@ -425,35 +437,184 @@ def _create_cells_and_feeds(self):
+                gid_srcs = net.gid_dict[_long_name(src_type)]
+                if unique:
+                    gid_srcs = [gid_target + net.gid_dict[src_type][0]]
+                for gid_src in gid_srcs:
```
For a unique feed, it looks almost looks like each source gid is getting connected to *all* cells of the given type. Is that correct?

jasmainak replied:
nopes, for a unique feed, `gid_srcs` is a list containing *one* element. So you don't have a nested for loop. Instead you loop only over the target cells and `gid_src` is actually "inferred" from the `gid_target` rather than being read from `net.gid_dict` ... so yeah agree it is hackish. What I did is just a reorganization of the old logic. I think fixing this might be a great idea for another PR.

rythorpe replied:
I get it now. Thanks for the explanation!

rythorpe commented on hnn_core/network_builder.py:
```diff
@@ -425,35 +437,184 @@ def _create_cells_and_feeds(self):
+                                nc_dict)
+        nc_dict['lamtha'] = 20.
+        nc_dict['A_weight'] = params[f'gbar_L2Basket_{target_cell}']
+        self._connect_celltypes('L2Basket', target_cell, 'soma', 'gabaa',
+                                nc_dict)
```
Are autapses supposed to be allowed here but not for L5 basket cells below?

jasmainak replied:
it's what the old code says: https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/basket.py#L78

and:

looks commented in HNN code too: 

https://github.com/jonescompneurolab/hnn/blob/master/L2_basket.py#L88

jasmainak replied:
I'm glad to see that cleaning up the code is unearthing these issues which will obviously lead to more consistent and better simulations

rythorpe replied:
Sure enough. @stephanie-r-jones do you know why L2 basket cells are allowed to synapse on themselves but the same isn't true for L5 basket cells?

rythorpe commented on hnn_core/tests/test_network.py:
```diff
@@ -54,6 +54,25 @@ def test_network():
+    # proximal
+    assert 'L2Pyr_L2Pyr_nmda' in network_builder.ncs
+    n_pyr = len(net.gid_dict['L2_pyramidal'])
+    n_connections = 3 * (n_pyr ** 2 - n_pyr)
```
I'm probably missing something obvious, but why is there a factor of 3 here?

jasmainak replied:
it's because there are 3 synapses per cell which are connected for NMDA.

rythorpe replied:
Oh that makes so much sense!

jasmainak commented:
@blakecaldwell SOS SOS SOS :) How do I reproduce the MPI bug locally? Are you able to reproduce it? I suspect something is wrong with `_PC.gid_exists`.

I tried activating the `testenv` on my laptop but tests pass. Also ... how do you debug stuff? Do you put print messages? This doesn't seem very optimal but I can live with it if it's at least on my laptop rather than a remote computer ...

Do you have a bit of time tomorrow to look at it together with me? I tried to ssh into Circle and debugging but it's super painful. Would be good to sort this out before the sprint since this PR would pave the path for some of the work that @rythorpe might be doing.

jasmainak commented:
okay I got MPI to work, yohooo 🥳 🥳 🥳 

Also addressed your comments @rythorpe with additional [documentation](https://github.com/jonescompneurolab/hnn-core/pull/164/files#diff-248a9986cd9a62479f5f12b0b576c072R460-R463) and [tests](https://github.com/jonescompneurolab/hnn-core/pull/164/files#diff-24892ee375b8aef5530b557e283ced19R76-R81). I'll open a couple of issues that came up and can perhaps be addressed by someone wanting to get started with the `Network` class

jasmainak commented:
It would be good to have this merged before the coding sprint so people don't have to deal with rebasing nightmares. Sorry for the pressure ... 🤐 🤐 🤐

rythorpe commented on hnn_core/network_builder.py:
```diff
@@ -425,35 +437,189 @@ def _create_cells_and_feeds(self):
+        # XXX: self.cells and net._gid_list are not same length
+        # ideally self.cells should be a dict of list
+        for gid_target, target_cell in zip(self.net._gid_list, self.cells):
+            is_target_gid = (gid_target in
+                             self.net.gid_dict[_long_name(target_type)])
+            if _PC.gid_exists(gid_target) and is_target_gid:
```
So the MPI bug was caused by referencing a gid that was not assigned to the current node within a parallel context?

jasmainak replied:
It's a bit hard to explain and I don't fully understand it, but I can assure you that it works fine :)

The problem is that there isn't necessarily a one-to-one mapping from `self.cells` to gids. That is `self.cells[175]` is not the cell corresponding to `gid=175`. Rather the one-to-one corresponding is between `self.cells` and `self._gid_list`. [This line](https://github.com/jonescompneurolab/hnn-core/pull/164/commits/8d3e5e7269b7e6a20c1298e21b4c6b1c8a62142d#diff-248a9986cd9a62479f5f12b0b576c072L479) was failing. What's not clear to me is why it was working in the non-MPI case ...

rythorpe replied:
I trust that it works - I'm just curious at this point. It looks like `self.cells` is [assigned according to the contents of `self._gid_list`](https://github.com/jonescompneurolab/hnn-core/blob/d2addbdea81ad73cab1e72371defbd89cf827a35/hnn_core/network_builder.py#L391). Since the contents of `self._gid_list` is [distinct for each node](https://github.com/jonescompneurolab/hnn-core/blob/d2addbdea81ad73cab1e72371defbd89cf827a35/hnn_core/network_builder.py#L332-L334), I guess it makes sense that `MPIBackend` (which parallelizes across cells) uniquely maps distinct gids and their corresponding cells to different nodes.

jasmainak replied:
yeah, I wonder if `self.net._gid_list` should be a `dict` called `gid_dict_node`, so it's easier to manipulate and more transparent what it does

rythorpe reviewed (approved):
This looks really good @jasmainak!
